timefold-solver/src/main/java/com/baeldung/timefoldsolver/ShiftScheduleApp.java

```diff
@@ -0,0 +1,54 @@
+package com.baeldung.timefoldsolver;
+
+import java.time.Duration;
+import java.time.LocalDate;
+import java.util.List;
+import java.util.Set;
+
+import ai.timefold.solver.core.api.solver.Solver;
+import ai.timefold.solver.core.api.solver.SolverFactory;
+import ai.timefold.solver.core.config.solver.SolverConfig;
+
+public class ShiftScheduleApp {
+
+    public static void main(String[] args) {
```

Comment:
Let's replace this main method with a unit test, I don't think we encourage main methods in examples, and prefer tests so we have a way to detect regressions in future

Reply:
If we replace this with a unit test as it, it will take 5 seconds from every CI build.

We should probably avoid that by changing

`.withTerminationSpentLimit(Duration.ofSeconds(5))`

to

`.withTerminationConfig(new TerminationConfig().withBestScoreLimit("0hard/0soft"))`

that will bring it down to 74milliseconds on my machine.

Reply:
In the article, I recommend against making that change. This only works if we know the score that the dataset will converge too (which is normally a catch 22 situation).

Reply:
I have not made this change yet. Let me know what you prefer :)

Reply:
> In the article, I recommend against making that change. 

Where is that in the article sorry? I missed it

And for my understanding, are we saying that `withTerminationSpentLimit` is more like "runs for exactly this amount of time" rather than "runs for at most this amount of time, exiting early if the most optimal solution has been found"?

It's not clear to me why it should take the full 5 seconds for such a small dataset, surely even a brute force approach would complete in less than this time? What am I missing here, does the solver do something monte-carlo like and not realise if it has exhausted all the possibilities already?

Also can you explain what `withBestScoreLimit` means, is this like putting an "exit early" condition in, when we reach some acceptable target metric threshold? 

Sorry for all the questions! I intentionally did not look through the documentation, I wanted fresh eyes to help come up with the kinds of questions a reader new to this might too :)

Reply:
Very good point, I 've brought it down to 2 second and added this in the article which is a practical get started approach to that complexity:

> We also need to configure how long we want a solver to run. For large datasets, with thousands of shifts and far more constraints, it's impossible to find the optimal solution in reasonable time (due to the exponential nature of [NP-hard](https://en.wikipedia.org/wiki/NP-hardness) problems). Instead, we want to find the best possible solution in the amount of time available. Let's limit that to 2 seconds for now:

I think a discussion of brute force scalability, NP-hard theory, etc is out of scope for this article.

Reply:
We do have a 3 hour video explaining that, in case we want to link that in the article :)
  https://youtu.be/FLaYKlLWKW4?si=GavSGoPIZ_dl91Vy&t=617
(currently not linked)

Reply:
> Very good point, I 've brought it down to 2 second and added this in the article which is a practical get started approach to that complexity:

Makes sense, looks good

> I think a discussion of brute force scalability, NP-hard theory, etc is out of scope for this article.

Agreed!

> We do have a 3 hour video explaining that, in case we want to link that in the article :)

Got it, will keep it in mind and let you know